### PR TITLE
Allow non-replicated ALTER TABLE FETCH/ATTACH in Replicated databases

### PR DESCRIPTION
**Pull request by kmichel-aiven**
`ALTER TABLE ... FETCH` and `ALTER TABLE ... ATTACH` queries were disabled in the Replicated database engine, because it could cause accidental duplication of data.

This enables these queries but without replicating them.

In the case of `FETCH`, the part will only be fetched on the server where the query is issued.

Similarly, in the case of `ATTACH`, the attached part only needs to be available on the server where the query is issued.

If the table itself is using one of the Replicated MergeTree engines, the attached data is then replicated by the table engine itself, without intervention of the database engine.

This change is meant to help with live backup/restore when using the Replicated database engine, using FREEZE for backup and ATTACH for restore.

**Changelog category (leave one):**
- New Feature

**Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):**
Allow non-replicated ALTER TABLE FETCH and ATTACH in Replicated databases

**Detailed description / Documentation draft:**
Included in the commit

---
```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

File: src/Interpreters/InterpreterAlterQuery.cpp [hunk 1]
```diff
@@ -55,7 +55,8 @@ BlockIO InterpreterAlterQuery::execute()
 
     DatabasePtr database = DatabaseCatalog::instance().getDatabase(table_id.database_name);
     if (typeid_cast<DatabaseReplicated *>(database.get())
-        && !getContext()->getClientInfo().is_replicated_database_internal)
+        && !getContext()->getClientInfo().is_replicated_database_internal
+        && !alter.isAttachAlter())
```
Comment (tavplubix) [inline src/Interpreters/InterpreterAlterQuery.cpp:59]: Btw, should we do the same for [ALTER FETCH PARTITION](https://github.com/ClickHouse/ClickHouse/blob/2dc5c89b66cff541bb66ee16203b87508e870572/src/Interpreters/executeDDLQueryOnCluster.cpp#L42)?

**Comment by tavplubix** [comment]
Also it would be great to remove `no-replicated-database` from some functional tests like [this](https://github.com/ClickHouse/ClickHouse/blob/39d73c01b225d5b274c11f75d786b6c81a53f1c8/tests/queries/0_stateless/01015_attach_part.sql#L2)

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Nice catch! You're right, we definitely should not replicate `DROP/DETACH/DROP DETACHED PART`, because part may not exist on some hosts, so query will never succeed and replication will hang in endless loop of retries. We can replicate `DROP/DETACH/FREEZE/DROP DETACHED PARTITION`, because these queries does nothing if specified partition does not exist. 

However, `DROP PARTITION` is usually used to remove some old data on the whole cluster (so it would be convenient to replicate it), but `DETACH PARTITION` + `ATTACH PARTITION` is usually used to reset `ReplicatedMergeTree` state in some partition and, for example, cancel merges on some shard (so we should not replicate it across shards). But it may be a bit confusion, so I think we can just disable replication of all such queries for consistency.

Another option is to introduce some setting that controls replication of such queries or some special syntax (like `ALTER ... [ON SHARD 'shard_name'] [ON REPLICA 'replica_name']`). But I'm not sure if it's a good idea, so let's just disable replication of `DROP/DETACH/DROP DETACHED PART/PARTITION` for now.

> this changes the behavior of an operation that was already allowed

It's ok, because Replicated database engine is experimental and we can break backward compatibility.

**Comment by kmichel-aiven** [comment]
Hi, I've seen the CONTRIBUTING file was edited, is the new CLA from the bot required or can we keep using the DCO ?

**Comment by tavplubix** [comment]
> Hi, I've seen the CONTRIBUTING file was edited, is the new CLA from the bot required or can we keep using the DCO ?

Sorry it took so long to get back to this PR. Yes, after recent changes new CLA is required.

**Review by tavplubix — APPROVED** [review]

**Pull request merged**